### PR TITLE
Convert PlayerHolder to Singleton; cleanup in VideoDetailFragment; Player/MainPlayer do not call onDestroy() directly

### PR DESCRIPTION
**Pull request by evermind-zz**
#### What is it?
- [ ] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

- PlayerHolder: convert to Singleton, handle context within, prevent ServiceConnection leak
- VideoDetailFragment: use viewBinding correctly
- VideoDetailFragment: remove duplicated code for toggle Fullscreen
- VideoDetailFragment: move null checks for player and playerService to helper methods
- Player/MainPlayer: service.onDestroy() should only be called from the system and not manually

#### APK testing

On the website the APK can be found by going to the "Checks" tab below the title and then on "artifacts" on the right.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java [hunk 1]
```diff
@@ -22,18 +23,27 @@
...
+    private static PlayerHolder instance;
+    public static synchronized PlayerHolder getInstance() {
+        if (PlayerHolder.instance == null) {
+            PlayerHolder.instance = new PlayerHolder();
+        }
+        return PlayerHolder.instance;
+    }
+
+    private final boolean DEBUG = MainActivity.DEBUG;
+    private final String TAG = "PlayerHolder";
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java:39]:
```suggestion
    private final String TAG = PlayerHolder.class.getSimpleName();
```

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java [hunk 2]
```diff
@@ -69,14 +83,19 @@ public static void setListener(final PlayerServiceExtendedEventListener newListe
         }
     }
 
-    public static void removeListener() {
+    public void removeListener() {
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java:86]: Remove this and let `setListener` receive a `@Nullable` parameter. When the parameter is `null` the listener should just be removed.

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java [hunk 3]
```diff
@@ -85,58 +104,65 @@ public static void startService(final Context context,
...
-            public void onServiceDisconnected(final ComponentName compName) {
-                if (DEBUG) {
-                    Log.d(TAG, "Player service is disconnected");
-                }
+    class PlayerServiceConnection implements ServiceConnection {
+
+        private boolean playAfterConnect = false;
+
+        public void doPlayAfterConnect(final boolean playAfterConnection) {
+            this.playAfterConnect = playAfterConnection;
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java:123]:
```suggestion
        public void doPlayAfterConnect(final boolean playAfterConnect) {
            this.playAfterConnect = playAfterConnect;
```

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 4]
```diff
@@ -2059,7 +2064,7 @@ private StackItem findQueueInStack(final PlayQueue queue) {
     }
 
     private void replaceQueueIfUserConfirms(final Runnable onAllow) {
-        @Nullable final PlayQueue activeQueue = player == null ? null : player.getPlayQueue();
+        @Nullable final PlayQueue activeQueue = !isPlayerAvailable() ? null : player.getPlayQueue();
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:2067]:
```suggestion
        @Nullable final PlayQueue activeQueue = isPlayerAvailable() ? player.getPlayQueue() : null;
```

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 5]
```diff
@@ -1059,6 +1065,14 @@ public void scrollToTop() {
     // Play Utils
     //////////////////////////////////////////////////////////////////////////*/
 
+    private void toggleFullscreenIfInFullscreenMode() {
+        //  If a user watched video inside fullscreen mode and than chose another player
+        //  return to non-fullscreen mode
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:1070]:
```suggestion
        // If a user watched video inside fullscreen mode and than chose another player
        // return to non-fullscreen mode
```

**Review by Stypox — CHANGES_REQUESTED** [review]
Looks mostly good. Thank you!
I also tested a little bit on API 29 and the player worked flawlessly; I read the logcat logs but found no strage errors, not even on player closing, so I think this changes are ok.

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java (hunk 3 above)
Comment (evermind-zz) [reply app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java:123]: This one was strange! I had it just like you suggested but checkStyle did not want to let me go. I'll recheck again

Comment (Stypox) [reply app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java:123]: Oh, then leave it as it is, nevermind

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java (hunk 2 above)
Comment (evermind-zz) [reply app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java:86]: you want me to remove the method removeListener() completly and replace it with a call to setListener(null);
Is that what you meant?

Comment (Stypox) [reply app/src/main/java/org/schabi/newpipe/player/helper/PlayerHolder.java:86]: Yes

**Comment by evermind-zz** [comment]
I've merged the changes and updated the pull request

**Review by Stypox — APPROVED** [review]
I also tested on API 19, everything works fine, thank you :-)

**Pull request merged**